main.rb

```diff
@@ -0,0 +1,36 @@
+require_relative 'lib/gateway/file'
+require_relative 'lib/use_case/count_views'
+
+require_relative 'lib/strategy/standard_count'
+require_relative 'lib/strategy/unique_count'
+require_relative 'lib/strategy/descending_order'
+
+require_relative 'lib/presenter/printable_list'
+
+file_gateway = Gateway::File.new(path: ARGV[0])
+
+all_descending = UseCase::CountViews.new(
+  file_gateway: file_gateway,
+  count_strategy: Strategy::StandardCount,
+  order_strategy: Strategy::DescendingOrder
+)
+
+unique_descending = UseCase::CountViews.new(
+  file_gateway: file_gateway,
+  count_strategy: Strategy::UniqueCount,
+  order_strategy: Strategy::DescendingOrder
+)
+
+puts "=" * 20
+puts "ALL VISITS"
+puts "=" * 20
+puts all_descending.execute(
+  presenter: Presenter::PrintableList.new(title: 'visits')
+)
+
+puts "=" * 20
+puts "UNIQUE VIEWS"
+puts "=" * 20
+puts unique_descending.execute(
+  presenter: Presenter::PrintableList.new(title: 'unique visits')
+)
```

Comment:
There's a quite a lot of orchestration here and whole `main.rb` is breaking the dependency inversion principle (DIP, or D in SOLID) which IRL we could mitigate with an use of factories. One good gem for them in Ruby-land is called `dry-container` – though just like `main.cpp` in C++ can never be fully inverted, our factories would not be as well... but at least this orchestration would look less messy than it does here.

Reply:
https://speakerdeck.com/naliwajek/use-case-factory-learn-tech-fridays